File: R/eln.R
logsumexp <- function(exponents.vec){
  m <- max(exponents.vec)
  m + log(sum(exp(exponents.vec-m)))
}

elnsum <- function(elnx, elny){
  ifelse(
    elnx == -Inf, elny, 
  ifelse(
    elny == -Inf, elnx,
  ifelse(
    elny < elnx,
    elnx + log(1+exp(elny-elnx)),
    elny + log(1+exp(elnx-elny)))))
}

elnproduct <- function(elnx, elny){
  ifelse(
    elnx == -Inf | elny == -Inf,
    -Inf,
    elnx+elny)
}

File: src/eln.cpp
#include <math.h>//for INFINITY.

// http://bozeman.genome.washington.edu/compbio/mbt599_2006/hmm_scaling_revised.pdf

double elnsum(double elnx, double elny){
  if(elnx == -INFINITY){
    return elny;
  }else if(elny == -INFINITY){
    return elnx;
  }else if(elny < elnx){
    return elnx + log(1+exp(elny-elnx));
  }else{
    return elny + log(1+exp(elnx-elny));
  }
}

double elnproduct(double elnx, double elny){
  if(elnx == -INFINITY || elny == -INFINITY){
    return -INFINITY;
  }else{
    return elnx+elny;
  }
}
